nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,378 @@
+	id := d.Id()
+	err := prepareForUpgrade(d, m)
+	if err != nil {
+		return handleUpdateError("NsxtUpgradePrepare", id, err)
+	}
+	return resourceNsxtUpgradePrepareRead(d, m)
+}
+
+func resourceNsxtUpgradePrepareDelete(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+func uploadPrecheckAndUpgradeBundle(d *schema.ResourceData, m interface{}) error {
+	upgradeBundleURL := d.Get("upgrade_bundle_url").(string)
+	precheckBundleURL := d.Get("precheck_bundle_url").(string)
+	upgradeBundleType := nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_UPGRADE
+	precheckBundleType := nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_PRE_UPGRADE
+	version := d.Get("version").(string)
+	c := m.(nsxtClients)
+	userName := c.NsxtClientConfig.UserName
+	password := c.NsxtClientConfig.Password
+	if !precheckBundleCompatibilityCheck(precheckBundleURL) {
+		return fmt.Errorf("Precheck bundle is only supported and is required for NSXT version >= 4.1.1")
+	}
+	if len(precheckBundleURL) > 0 {
+		err := uploadUpgradeBundle(m, precheckBundleURL, userName, password, precheckBundleType, version)
+		if err != nil {
+			return fmt.Errorf("Failed to upload precheck bundle: %s", err)
+		}
+	}
+	err := uploadUpgradeBundle(m, upgradeBundleURL, userName, password, upgradeBundleType, version)
+	if err != nil {
+		return fmt.Errorf("Failed to upload upgrade bundle: %s", err)
+	}
+	return nil
+}
+
+func precheckBundleCompatibilityCheck(precheckBundleURL string) bool {
+	if nsxVersionLower("4.1.1") && len(precheckBundleURL) > 0 {
+		return false
+	}
+	if nsxVersionHigherOrEqual("4.1.1") && len(precheckBundleURL) == 0 {
+		return false
+	}
+	return true
+}
+
+func uploadUpgradeBundle(m interface{}, url string, userName string, password string, bundleType string, version string) error {
+	connector := getPolicyConnector(m)
+	client := upgrade.NewBundlesClient(connector)
+	bundleFetchRequest := nsxModel.UpgradeBundleFetchRequest{
+		Url: &url,
+	}
+	if nsxVersionHigherOrEqual("4.1.1") {
+		bundleFetchRequest.BundleType = &bundleType
+		bundleFetchRequest.Password = &password
+		bundleFetchRequest.Username = &userName
+		bundleFetchRequest.Version = &version
+	}
+	bundleID, err := client.Create(bundleFetchRequest, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to upload upgrade bundle of type %s: %v", bundleType, err)
+	}
+	return waitForBundleUpload(m, *bundleID.BundleId)
+}
+
+func acceptUserAgreement(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	acceptUserAgreement := d.Get("accept_user_agreement").(bool)
+	if !acceptUserAgreement {
+		return fmt.Errorf("To proceed with upgrade, you must accept user agreement")
+	}
+	client := eula.NewAcceptClient(connector)
+	err := client.Create()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeUc(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewUpgradeClient(connector)
```

Comment:
Same comment as above - should we upgrade UC if this step was already executed succesfully?

Reply:
Done